core/api/src/main/scala/org/virtuslab/ideprobe/jsonrpc/logging/RequestResponseLogger.scala

```diff
@@ -0,0 +1,76 @@
+package org.virtuslab.ideprobe.jsonrpc.logging
+
+import java.time.ZonedDateTime
+import java.util.concurrent.{ScheduledFuture, ScheduledThreadPoolExecutor, TimeUnit}
+
+class RequestResponseLogger {
+
+  private val ex = new ScheduledThreadPoolExecutor(1)
+  private val ticker: Runnable = () => flush()
+  private var task: Option[ScheduledFuture[_]] = None
+
+  private sealed trait LogEntry
+  private case class Request(message: String) extends LogEntry
+  private case class Response(message: String) extends LogEntry
+  private case class RequestAndResponse(request: String, response: String) extends LogEntry
+
+  private var buffer: List[(LogEntry, ZonedDateTime)] = List.empty
+
+  def logRequest(request: String): Unit = buffer.synchronized {
+    task.foreach(_.cancel(true))
+    buffer match {
+      case (head @ (RequestAndResponse(`request`, _), _)) :: tail =>
+        buffer = (Request(request), ZonedDateTime.now()) :: head :: tail
+      case _ =>
+        flush()
+        buffer = (Request(request), ZonedDateTime.now()) :: Nil
+    }
+    task = Option(ex.schedule(ticker, 1, TimeUnit.SECONDS))
```

Comment:
let's extract this 1 second to a val of type scala.concurrent.Duration. We might further put it in configuration

Reply:
Done.